Assets/Scripts/Player/PlayerController.cs

```diff
@@ -1,93 +1,133 @@
+        nextFire = Time.time;
     }
 
     public int GetMaxHealth() => maxHealth;
     public int GetCurrentHealth() => currentHealth;
 
-    private void Update() {
-        if (PauseController.Instance.GamePaused) {
-            return;
-        }
+    private void Update()
+    {
+        if (PauseController.Instance.GamePaused) { return; }
 
-        if (Input.touchCount > 0) {
+        if (Input.touchCount > 0)
+        {
             Touch touch = Input.GetTouch(0);
 
-            if (touch.phase == TouchPhase.Began && !EventSystem.current.IsPointerOverGameObject(touch.fingerId)) {
+            if (touch.phase == TouchPhase.Began && !EventSystem.current.IsPointerOverGameObject(touch.fingerId))
+            {
                 targetPosition = Camera.main.ScreenToWorldPoint(touch.position);
             }
         }
-        else if (Input.GetMouseButtonDown(0) && !EventSystem.current.IsPointerOverGameObject()) {
+        else if (Input.GetMouseButtonDown(0) && !EventSystem.current.IsPointerOverGameObject())
+        {
             targetPosition = Camera.main.ScreenToWorldPoint(Input.mousePosition);
         }
 
-        if (Input.GetKeyDown(KeyCode.H))
-        {
-            Heal(20);
-        }
+        if (rb2D.position == targetPosition && enemyHash.Count > 0) { RangedAttack(); }
+
+        if (Input.GetKeyDown(KeyCode.H)) { Heal(20); }
+
+        if (Input.GetKeyDown(KeyCode.Space)) { TakeDamage(20); }
 
-        if (Input.GetKeyDown(KeyCode.Space))
-        {
-            TakeDamage(20);
-        }
         if (GetCurrentHealth() <= 0)
         {
-            
             Scene currentScene = SceneManager.GetActiveScene();
             SceneManager.LoadScene(currentScene.name);
-          
         }
     }
 
-    private void FixedUpdate() {
+    private void FixedUpdate()
+    {
         Vector2 newPosition = Vector2.MoveTowards(rb2D.position, targetPosition, movementSpeed);
         rb2D.MovePosition(newPosition);
 
         // TODO: Maybe set targetPosition to currentPosition when the player is stopped. Then, the player
         // won't continually try to reach an unreachable targetPosition when blocked by a collider.
     }
 
-    public void TakeDamage(int damage)
+    private void RangedAttack()
     {
+        if (Time.time > nextFire)
+        {
+            Instantiate(projectile, transform.position, Quaternion.identity);
+            proj = projectile.GetComponent<ProjectileController>();
 
-        currentHealth -= damage;
-        healthBar.SetHealth(currentHealth);
+            if (proj != null)
+            {
+                proj.target = FindTargetEnemy();
+                proj.senderColl = playerColl;
+            }
+            nextFire = Time.time + fireRate;
+        }
+    }
 
-        if (currentHealth < 0)
+    private Transform FindTargetEnemy()
+    {
+        Transform closeEnem = null;
+        float distToCloseEnem = Mathf.Infinity,
+              distToEnem = 0;

```

Comment:
I think that `distToEnem` should be declared inside of the `foreach` loop because it doesn't need to be accessible outside of that loop. I don't think there's much of a benefit either way, though.

Reply:
My thought process is creating those variables one per function call is more efficient than once per enemy, but I also recognize that the function is running when there are no enemies in the hashset so I just added to the if condition on line 88 to prevent that.

Reply:
Efficiency is not meaningfully affected either way. I think it's better for the variable to have a smaller scope in this case, but you can leave it as is if you want. When variables have a smaller scope, it's easier to see where variables are being affected and extract code into a separate method if needed. In this case, the scope change doesn't really make much of a difference.